### ISSUE TITLE
xASL_stat_GetROIstatistics - bug with x.S.SUBJECTID if last subject_run does not have ASL data

### ISSUE THREAD

**Issue by MauricePasternak** [ExploreASL/ExploreASL#105]
To recreate the bug, run any dataset under the following conditions:
- x.SkipIfNoASL = 1
- The very last subject/run must not have an ASL scan

If you breakpoint at line 461 in xASL_stat_GetROIstatistics and iterate through the loops, you will observe that x.S.SUBJECTID either adds the appropriate name (ex 'sub_001_ASL_1') or [ ]. However, if the final subject/run did not have an ASL scan, a [ ] will not be placed in the nth cell. This causes a fatal error later on line 127 in xASL_stat_PrintStats because iSubjectSession will always be greater than the number of cells in x.S.SUBJECTID.

Found by accident while running the 22q11 dataset where subject 027 (of 27) does not have an ASL scan.

**Comment by MauricePasternak** [comment]
![issue105 - LastSubject_Run has no ASL bug - MP Additions](https://user-images.githubusercontent.com/57411571/91082771-2786bf80-e617-11ea-891f-78c7d23cbb26.png)
![issue105 - LastSubject_Run has no ASL bug - without MP Additions](https://user-images.githubusercontent.com/57411571/91082774-28b7ec80-e617-11ea-951c-8f9c193a0117.png)
![issue105 - LastSubject_Run has no ASL bug - DAT not updated](https://user-images.githubusercontent.com/57411571/91082779-29e91980-e617-11ea-920c-419d6cc86ef0.png)

This issue may stem a little deeper because it is not just x.S.SUBJECTID whose array length was not updated. Other structs seem to be affected too, such as x.S.DAT. I've attached images showing the issue and a solution for x.S.SUBJECTID, but the other arrays will also have to be made aware when the last subject_run is skipped over. x.S.DAT, for example, should have 0, 0, 0 on row 27, but instead it is glanced over. We will likely have to check what other arrays are impacted similarly.

**Comment by HenkMutsaerts** [comment]
Can you save the working memory of Matlab in a .mat file, at the line in a function where this goes wrong? That makes it much easier for me to fix this! Thanks!

**Comment by HenkMutsaerts** [comment]
## Generalization
If I generalize your issue, it means that empty results are not always filled with NaNs. If a subject or session was not included at the beginning of the pipeline, it should be skipped at all times (since this can be on purpose). If subjects and/or sessions were included in the pipeline, but are missing, for whatever reason, we print their name/session IDs, but fill their data with NaNs.

## xASL_stat_PrintStats: issue warning if incomplete data, instead of crashing.
This function just prints, so shouldn't recreate missing data but rather issuing a warning instead of crashing.

## xASL_stat_GetROIStatistics: now adds lines with NaNs for missing data.
This function should now deal with the missing data:
* Start with initializing x.S.SUBJECTID
* Always initialize x.S.DAT fields
* Finalize by checking if last row is missing, and create it with NaNs if it is missing
It remains annoying that Matlab fills missing numerical data with zeros instead of NaNs, but this should be solved by predefining matrices with NaNs.

**Comment by MauricePasternak** [comment]
I will re-test with a larger sample size, but from the mini-batch of 4 subjects (2 with ASL and 2 without), the pipeline fails due to the following condition in xASL_wrp_CreatePopulationTemplates starting @ line 108:

```
if x.nSubjectsSessions<6
    % With too small datasets, created templated won't be reliable
    fprintf('%s\n',['Too few images (' num2str(x.nSubjectsSessions) ') for creating templates/parametric images, skipping...']);
    return;
end

Size1 = sum(x.WBmask(:));

x.D.TemplatesStudyDir = fullfile(x.D.PopDir, 'Templates');
xASL_adm_CreateDir(x.D.TemplatesStudyDir);
```

So the snowball effect is this:

1) analysis/Population/Templates is never created (due to early cancellation at low subject number)

2) various filepath variables in xASL_im_CreateAnalysisMask have the value: ""  (an empty string) as a result.

3)  xASL_exist causes the crash when the input is:  xASL_exist("", 'file')

So, from my observations, we can either go with the philosophy of having xASL_exist have some sort of try-catch method. Or prevent the issue from happening by allowing the Templates directory to exist anyway and have the right files in there, but to forego the actual creation of the analysis mask later.

**Comment by HenkMutsaerts** [comment]
This test was with issue #143, and branch bug-#143_MasksMissing I assume?
I will move your comment there and answer it ;)

**Comment by jan-petr** [comment]
@HenkMutsaerts henk The recreating and cherry-picking didn't work well even with -S for the last commit that was conflicting. After finishing of the cherry-picking and conflict resolving, the last edit remained unsigned. Check if you can do something about it...

**Comment by MauricePasternak** [comment]
I'm a little late prior to the closure. But I did want to verify that, using the develop branch, crashes no longer occur. 

Disregard the jerry-rigged 345434 censoring. Just be aware that it looks like missing data for subject_runs that are NOT the last run will be imputed as zeros while a subject_run with missing data who happens to be the last subject_run will be imputed with n/a.

![Untitled](https://user-images.githubusercontent.com/57411571/94586541-59f78e00-024f-11eb-92e9-01445f2bd8bc.png)

This is just me splitting hairs, but if possible, I'd prefer n/a instead of zeros in all instances where data was missing.

**Comment by HenkMutsaerts** [comment]
Yes @MauricePasternak good point, I added [THIS](https://github.com/ExploreASL/ExploreASL/issues/171) issue to be fixed after release 1.3.0.
